### ISSUE TITLE
Duplicating URIs?

### ISSUE THREAD

**Issue by sakars** [uriparser/uriparser#200]
Hi!
I wish to use this library in a C++ project, by crafting a wrapper around it to make it more memory safe, however, while poking around, I couldn't find a good way of duplicating `UriUriA` objects. What is the proper way of performing a deep copy of `UriUriA`s?

A simple shallow clone, `owner=false`, `uriMakeOwnerA(&uri)` didn't do the trick and later member releases performed a double free.

**Comment by sakars** [comment]
The only way I think I'm getting this to work is to recompose the `UriUriA` and then re-parse it, but that seems like a waste and it's quite a bit of hoop jumping on my wrapper.

**Comment by hartwork** [comment]
Hi @sakars I understand your goal. I rechecked the function list at https://uriparser.github.io/doc/api/latest/globals_func.html#index_u to be sure I'm not missing a function but status quo seems indeed to be that the API does not allow duplicating `UriUriA` instances in a safe and moderately easy manner. Maybe you're the first in 15+ years trying to wrap uriparser in C++ classes. I understand it's a useful feature but I have a lot of other stuff on my plate, so I won't be able to provide short solutions I'm afraid.

**Comment by sakars** [comment]
Thank you for a swift response!

The reason I'm trying to use this library at all instead of any C++ libraries is because of lightweight dependencies, but more importantly, because it can perform reference resolution as defined in RFC3986 Section 5. Surprisingly few libraries implement that.

The C++ wrapper wouldn't be thorough, I'm just wrapping the functionality I need so I can use it without memory safety concerns. Might just have to bite the bullet and make my own if no libraries suit my needs.

**Comment by hartwork** [comment]
> Might just have to bite the bullet and make my own if no libraries suit my needs.

@sakars if you do go that route of making your own parser, please note that there is [**I**RI (RFC 398**7**)](https://datatracker.ietf.org/doc/html/rfc3987#section-1.1) by now, not just URI (RFC 3986).

**Comment by sakars** [comment]
Hey, quick update.

> @sakars if you do go that route of making your own parser, please note that there is [**I**RI (RFC 398**7**)](https://datatracker.ietf.org/doc/html/rfc3987#section-1.1) by now, not just URI (RFC 3986).

Ended up using uriparser.
Didn't even know IRI existed ^^; But that isn't what I need as I'm conforming to JSON Schema draft standards which only require URIs.

I rewrote my C++ classes and they now work reasonably well for my use cases.

If anyone ends up coming across this with a similar issue, I fixed moving by using `unique_ptr`s around `UriUriA`s, effectively not moving them in the first place, and copied URIs by reforming the URI to strings, then back into a URI. Haven't yet found any issues to the combination to these methods.
